### ISSUE TITLE
Technical: AI-Powered Conversation Summarization Backend

### ISSUE THREAD

**Issue by aponomy** [aponomy/senior-ai-mvp#2]
## Overview

Develop the backend AI and algorithmic systems to support the progressive conversation summarization interface, enabling intelligent conversation segmentation, summarization, and context management.

### 1. Conversation Beat Detection System

Develop an algorithm to automatically identify natural conversation boundaries and group messages into logical "beats" or sequences.

**Detection Signals**:
- **Topic Shift Detection**: Analyze semantic similarity between consecutive messages to detect topic changes
- **Intent Classification**: Identify when user switches from question → command → chitchat
- **Temporal Gaps**: Detect pauses in conversation (>2 minutes as threshold)
- **Task Completion Markers**: Recognize phrases like "thanks", "got it", "what about...", "now let's..."
- **Conversation Depth**: Track nested topics (subtopic within topic)

**Algorithm Requirements**:
- Real-time processing (new messages analyzed immediately)
- Sliding window analysis (look at 3-5 message context)
- Configurable sensitivity for grouping
- Output: Conversation segments with start/end indices and beat type

**Example Beat Types**:
```typescript
type BeatType = 
  | 'question-answer'    // User asks, AI answers
  | 'task-execution'      // User requests action, AI performs
  | 'clarification'       // Back-and-forth clarifying questions
  | 'chitchat'           // Social conversation
  | 'topic-exploration'  // Deep dive into subject
```

### 2. Multi-Level Summarization Engine

Create a summarization system that generates different levels of summary for the same conversation segment.

**Three Summary Levels**:

1. **L1 - Condensed (for semi-transparent boxes)**
   - 50-70% of original length
   - Remove filler words, redundancy
   - Keep key information intact
   - Example: "You asked about Italian recipes. I suggested three options: pasta carbonara, risotto, and lasagna."

2. **L2 - Abstract (for solid summary boxes)**
   - 15-25% of original length
   - High-level topic identification
   - One sentence typically
   - Example: "Discussion about Italian recipe options"

3. **L3 - Label (for very old conversations, future feature)**
   - 2-5 words
   - Minimal description
   - Example: "Italian Recipes"

**Technical Approach Options**:
- **Option A**: Use LLM API (OpenAI, Anthropic) with custom prompts per level
- **Option B**: Use local transformer model (fine-tuned summarization model)
- **Option C**: Hybrid - extractive summarization for L1, abstractive for L2
- **Option D**: Use embeddings + clustering + template generation

**Performance Requirements**:
- L1 summary: <500ms per beat
- L2 summary: <300ms per beat
- Batch processing for efficiency
- Cache summaries (regenerate only when conversation changes)

### 3. Dynamic Threshold Management

Implement adaptive logic to determine when messages transition from one visualization tier to another.

**Factors to Consider**:
- Total conversation length (more messages = more aggressive summarization)
- Conversation recency (active conversation vs dormant)
- User interaction patterns (if user keeps expanding, show more detail)
- Viewport size (smaller screens = more summarization)
- User preferences (power users might want less summarization)

**Example Threshold Logic**:
```typescript
interface ThresholdConfig {
  recentMessageCount: number;      // Show full text for last N messages
  nearRecentMessageCount: number;  // Show condensed for next N messages
  maxBeatsBeforeSummary: number;   // Max conversation beats before summarizing
  beatMessageThreshold: number;    // Min messages to form a beat
}

// Dynamic calculation based on conversation state
function calculateThresholds(
  totalMessages: number,
  conversationAge: number,
  userInteraction: InteractionMetrics
): ThresholdConfig
```

### 4. Conversation State Management

Build a robust state management system to track conversation structure, summaries, and user interactions.

**State Structure**:
```typescript
interface ConversationState {
  messages: Message[];
  beats: ConversationBeat[];
  summaries: Map<beatId, SummaryLevels>;
  expansionState: Map<beatId, ExpansionLevel>;
  metadata: ConversationMetadata;
}

interface ConversationBeat {
  id: string;
  messageIds: string[];
  startIndex: number;
  endIndex: number;
  beatType: BeatType;
  timestamp: Date;
  topic?: string;
}

interface Message {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: Date;
  beatId?: string;
  metadata?: any;
}

interface SummaryLevels {
  l1: string;  // Condensed
  l2: string;  // Abstract
  l3?: string; // Label (optional)
  generatedAt: Date;
}
```

**State Operations**:
- Add new message → Update beats → Generate summaries
- Expand beat → Load full messages → Track expansion state
- Collapse beat → Update UI state
- Navigate conversation → Load relevant summaries
- Persist state (localStorage/IndexedDB for offline support)

### 5. Context Window Management

Implement smart context management for the AI assistant to handle long conversations efficiently.

**Challenges**:
- AI models have token limits (e.g., 128k tokens)
- Need to maintain conversation coherence
- Must provide relevant context for new responses

**Strategy**:
- **Recent Context**: Always include last 10-20 messages in full
- **Summarized Context**: Include L2 summaries of older beats
- **Semantic Search**: If user references old content, retrieve relevant beats
- **Context Injection**: Dynamically build context window based on current topic

**Example Context Building**:
```typescript
function buildAIContext(
  state: ConversationState,
  currentTopic: string
): AIContext {
  return {
    recentMessages: state.messages.slice(-20),
    relevantBeats: findRelevantBeats(state, currentTopic),
    conversationSummary: generateOverallSummary(state),
  };
}
```

### 6. Performance Optimization

Ensure the system runs smoothly even with very long conversations.

**Optimization Strategies**:
- **Lazy Loading**: Only load/render visible conversation segments
- **Virtual Scrolling**: Render only messages in viewport
- **Memoization**: Cache computed summaries and beat analyses
- **Web Workers**: Run beat detection and summarization in background
- **Progressive Enhancement**: Show skeleton UI while processing
- **Batch Operations**: Process multiple summaries in single API call

**Performance Targets**:
- New message processing: <100ms
- Beat detection: <200ms per message
- Summary generation: <500ms per beat
- UI rendering: 60fps during scroll
- Memory usage: <100MB for 1000 messages

### 7. Error Handling & Fallbacks

Build resilient system with graceful degradation.

**Error Scenarios**:
- Summarization API failure → Use extractive fallback
- Beat detection failure → Use time-based grouping
- Context too large → Aggressive summarization
- Network offline → Use cached summaries
- Invalid state → Rebuild from messages

**Fallback Hierarchy**:
1. Prefer AI-generated summaries
2. Fall back to extractive summarization
3. Fall back to time-based grouping
4. Fall back to simple message truncation

### System Components

```
┌─────────────────────────────────────────────────────────┐
│  Conversation Component (React)                          │
│  - Renders progressive UI                                │
│  - Handles user interactions                             │
└────────────────┬────────────────────────────────────────┘
                 │
┌────────────────▼────────────────────────────────────────┐
│  Conversation Manager (State)                            │
│  - Manages conversation state                            │
│  - Coordinates beat detection & summarization            │
└────────────┬───────────────────┬────────────────────────┘
             │                   │
┌────────────▼─────────┐  ┌──────▼──────────────────────┐
│  Beat Detector        │  │  Summarization Engine        │
│  - Topic analysis     │  │  - L1 condensed summaries    │
│  - Intent detection   │  │  - L2 abstract summaries     │
│  - Grouping logic     │  │  - Caching & optimization    │
└──────────────────────┘  └─────────────────────────────┘
```

### API Integration Points

**If using external LLM API**:
- Endpoint for beat detection (optional, or do locally)
- Endpoint for summarization (multiple levels)
- Streaming support for real-time summaries
- Rate limiting and queue management

**Local Processing**:
- Embedding generation (for topic similarity)
- Intent classification model
- Extractive summarization
- All running client-side for privacy

## Implementation Phases

**Phase 1: Foundation**
- Basic beat detection using time gaps and keyword triggers
- Simple extractive summarization
- State management setup
- In-memory storage

**Phase 2: AI Integration**
- Integrate LLM API for summarization
- Improve beat detection with semantic analysis
- Add caching layer
- Persistent storage

**Phase 3: Optimization**
- Web Worker for background processing
- Virtual scrolling implementation
- Performance profiling and optimization
- Memory management

**Phase 4: Advanced Features**
- Semantic search across conversation
- Context-aware summary regeneration
- User preference learning
- Export/import conversation state

## Technology Stack Recommendations

**Beat Detection**:
- TensorFlow.js for client-side ML (optional)
- Natural NLP library for basic text analysis
- Custom heuristics for pattern matching

**Summarization**:
- OpenAI GPT-4 API or Anthropic Claude API (cloud)
- OR: Hugging Face Transformers (local models)
- OR: Custom extractive algorithm (fallback)

**State Management**:
- Zustand or Redux Toolkit
- IndexedDB for persistence (via Dexie.js)
- React Query for API caching

**Performance**:
- React Virtual for virtual scrolling
- Comlink for Web Worker communication
- React.memo and useMemo for optimization

## Open Technical Questions

1. Should beat detection run on every message or batch process periodically?
2. Local ML models vs cloud API - what's the right balance?
3. How to handle conversation branching (if we add that feature)?
4. Optimal storage format for very long conversations?
5. Should we use vector database for semantic search?
6. How to handle multi-language conversations?
7. Privacy considerations - should summaries be generated locally?

## Success Criteria

- Beat detection accuracy >85% (human evaluation)
- Summary quality score >4/5 (user ratings)
- No perceptible lag when adding new messages
- Support for 1000+ message conversations
- <2 second cold start for loading conversation

---

## Labels

- `enhancement`
- `backend`
- `ai-ml`
- `performance`

## Priority

High - Required for core conversation interface feature (Issue #1)

**Comment by aponomy** [comment]
## Analysis: BeatType Design + Research Integration

After analyzing the proposed `BeatType` against previous conversation beat research (grounding theory, fixpoints, speech acts), here's the recommendation:

### ✅ Keep BeatType Simple (as proposed)

The current design is **perfect for elderly users**:
```typescript
type BeatType = 
  | 'question-answer'
  | 'task-execution'
  | 'clarification'
  | 'chitchat'
  | 'topic-exploration'
```

**Why this works:**
- User-facing and understandable (85-year-olds can grasp these categories)
- Purpose-driven (reflects user goals, not linguistic theory)
- Actionable for UI (clear what to show in summary boxes)
- Maintainable (won't confuse future developers)

### 🎯 Add ONE Critical Dimension: BeatStatus

Based on grounding theory research (communication fixpoints), add this:

```typescript
type BeatStatus = 
  | 'complete'           // Fixpoint reached, mutual understanding confirmed
  | 'incomplete'         // Hanging question, no resolution yet
  | 'tentative'          // Implicit acceptance, can still be reopened
```

### Why BeatStatus Adds Real UX Value:

1. **Visual Clarity**: Incomplete beats get subtle UI indicator (different border/icon)
2. **Context Recovery**: System knows "we never finished discussing recipes"
3. **Proactive Help**: AI can prompt: "Earlier you asked about X, but we didn't finish - continue?"
4. **Cognitive Support**: Helps users with memory concerns track resolved vs. pending topics
5. **Better Summarization**: Complete beats can be condensed more aggressively

### Updated ConversationBeat Structure:

```typescript
interface ConversationBeat {
  id: string;
  messageIds: string[];
  startIndex: number;
  endIndex: number;
  beatType: BeatType;           // WHAT happened (simple, user-facing)
  beatStatus: BeatStatus;       // Did we reach fixpoint? (adds value!)
  timestamp: Date;
  topic?: string;
  
  // Internal AI metadata (for decision-making, not UI):
  _internal?: {
    dominantMove?: 'add' | 'infer' | 'ask' | 'accept' | 'reject' | 'clarify';
    coordinationLevel?: 'basic' | 'description' | 'definition';
    confidenceScore?: number;
  }
}
```

### Detection Heuristics for BeatStatus:

**Complete:**
- User says "thanks", "got it", "perfect", "okay"
- AI asks "Does that help?" and user confirms
- Task successfully executed with confirmation
- Both parties move to new topic

**Incomplete:**
- User asks question, no response yet
- AI asks clarifying question, no answer
- Task failed or error occurred
- Conversation interrupted mid-flow

**Tentative:**
- AI provided answer, user didn't explicitly confirm but continued
- Information shared without explicit acceptance
- Implicit agreement (Clark's "grounding criterion")

### Mapping Research Concepts to Simple Implementation:

Your deep research **informs** the simple types without complicating them:

| Research Concept | Implementation |
|-----------------|----------------|
| BeatMove.Accept → | BeatStatus.complete |
| BeatMove.Reject + new topic → | BeatStatus.incomplete |
| BeatMove.Add without opposition → | BeatStatus.tentative |
| CoordinationLevel escalation → | BeatType 'clarification' |
| SpeechAct.Question → | BeatType 'question-answer' |
| Multiple RequestExpansion/Repair → | BeatType 'clarification' |

### Benefits of This Approach:

✅ **Simplicity** - Core types remain clean and understandable  
✅ **Richness** - AI has deep context for smart decisions  
✅ **Research-backed** - Fixpoint theory applied practically  
✅ **User value** - Incomplete conversations don't get lost  
✅ **Extensible** - Can add more internal metrics without breaking contract

### Implementation Notes:

1. **UI Indicators:**
   - Complete: Standard summary box
   - Incomplete: Box with dotted border + "Unfinished" badge
   - Tentative: Standard box, but can be "reopened" easily

2. **Context Window Management:**
   - Always include incomplete beats in AI context
   - Tentative beats can be summarized more aggressively
   - Complete beats are first candidates for deep summarization

3. **User Prompts:**
   - On session resume: "You have 2 incomplete conversations - would you like to continue?"
   - When detecting related topic: "This relates to our earlier incomplete discussion about X"

### Recommendation:

**Keep it simple, add status, win at UX.** The complex research concepts live in the AI's decision-making layer, not in the type system users/developers interact with.

This bridges theory with practice perfectly for an elderly-focused application. 🎯
